pesto/objective.py

```diff
@@ -1,30 +1,309 @@
+        dimension.
+
+    grad: callable, bool, optional
+        Method for computing the gradient vector. If it is a callable,
+        it should be of the form
+            ``grad(x) -> array_like, shape (n,).``
+        If its value is True, then fun should return the gradient as a second
+        output.
+
+    hess: callable, optional
+        Method for computing the Hessian matrix. If it is a callable,
+        it should be of the form
+            ``hess(x) -> array, shape (n,n).``
+        If its value is True, then fun should return the gradient as a
+        second, and the Hessian as a third output, and grad should be True as
+        well.
+
+    hessp: callable, optional
+        Method for computing the Hessian vector product, i.e.
+            ``hessp(x, v) -> array_like, shape (n,)``
+        computes the product H*v of the Hessian of fun at x with v.
+
+    res: {callable, bool}, optional
+        Method for computing residuals, i.e.
+            ``res(x) -> array_like, shape(m,).``
+
+    sres: callable, optional
+        Method for computing residual sensitivities. If its is a callable,
+        it should be of the form
+            ``sres(x) -> array, shape (m,n).``
+        If its value is True, then res should return the residual
+        sensitivities as a second output.
+
+    """
+
+    MODE_FUN = 'MODE_FUN'  # mode for function values
+    MODE_RES = 'MODE_RES'  # mode for residuals
+
+    def __init__(self, fun,
+                 grad=None, hess=None, hessp=None,
+                 res=None, sres=None):
         self.fun = fun
-        self.dim = dim
-        # dim rather in optimizer
+        self.grad = grad
+        self.hess = hess
+        self.hessp = hessp
+        self.res = res
+        self.sres = sres
+
+        """
+        TODO: 
+        
+        * Implement methods to compute grad via finite differences (with 
+        an automatic adaptation of the step size), 
+        and diverse approximations of the Hessian.
+        """
+
+    def __call__(self, x, sensi_orders: tuple, mode=MODE_FUN):
+        """
+        Method to get arbitrary sensitivities.
+
+        This method is a bit lengthy, because there are different ways in which
+        an optimizer calls the objective function, and in how the objective
+        function provides
+        information (e.g. derivatives via separate functions or along with
+        the function values). The different calling modes increase efficiency
+        in space and time.
+
+        Parameters
+        ----------
+
+        x: array_like
+            The parameters for which to evaluate the objective function.
+
+        sensi_orders: tuple
+            Specifying which sensitivities to compute, e.g. (0,1) -> fval, grad.
+
+        mode: str
+            Whether to compute function values or residuals.
+        """
+
+        if mode == Objective.MODE_RES:
+            if sensi_orders == (0,):
+                if self.sres is True:
```

Comment:
sres can be a callable, or a bool (indicating that res also returns sensitivities). So one needs to explicitly check for equality with True, because if object in python also evaluates to True, afaio.

Reply:
Got it.